File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        memo={}
        def dfs(i,total):
            if (i,total) in memo:
                return memo[(i,total)]
            if i>len(nums)-1:
                memo[(i,total)]=total
                return total
            memo[(i,total)]=max(dfs(i+1,total),dfs(i+2,total+nums[i]))
            return memo[(i,total)]
            
        
        return dfs(0,0)

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-5.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map=new HashMap<>();
        map.put(0,1);
        int total=0;
        int cnt=0;
        for(int num:nums){
            total+=num;
            int opposite=total-k;
            if(map.containsKey(opposite)){
                cnt+=map.get(opposite);
            }
            map.merge(total,1,Integer::sum);
        }
        return cnt;
    }
}